Backend + Improvement: Forge Key Events

rueblimaster opened the pull request:
## What
Before we just used the last Keyboard.Event each tick to read our inputs. Now we're connected to the normal way of doing this. We use forge events, that means on KeyInputEvent we go and check the keyboard for the event, this is done for all events, tho it's controlled by forge. Same principal applies to the mouse.

### Tested cases
- Keyboard keybinds
- Mouse keybinds
- Without a mouse plugged in

## Changelog Improvements
+ Keybinds now work when pressing multiple keys at the same time. - rueblimaster

## Changelog Technical Details
+ Switched over to forge events to power KeyEvents. - rueblimaster

rueblimaster committed:
Create Keybinding.kt

rueblimaster committed:
improve keybindings

rueblimaster committed:
made conditions nullable

rueblimaster committed:
massively improve keybindingType etc.

rueblimaster committed:
small cleanup / improvements

rueblimaster committed:
add questionable command

rueblimaster committed:
better names

rueblimaster committed:
use fancy new HandleEvent annotation

rueblimaster committed:
Revert "use fancy new HandleEvent annotation"

This reverts commit 40dc592fb06dd2877d70f1ffb9ac9e4fbe993258.

rueblimaster committed:
add back DebugDataCollect

rueblimaster committed:
fix wrong keybindingType

rueblimaster committed:
some detekt

rueblimaster committed:
use fancy new HandleEvent annotation

rueblimaster committed:
remove activeKeybindings

as the performance gain is only minor and not worth it

rueblimaster committed:
improve DebugDataCollect

rueblimaster committed:
remove keybindingType caching

rueblimaster committed:
implement in FocusMode

rueblimaster committed:
implement PestFinder

rueblimaster committed:
implement in GardenWarpCommands

rueblimaster committed:
implement in FishingTimer

rueblimaster committed:
detekt

rueblimaster committed:
add option requireSeparateTaps

rueblimaster committed:
detekt

rueblimaster committed:
Merge branch 'beta' into KeyBindUtils

rueblimaster committed:
names?

rueblimaster committed:
it, I hope

rueblimaster committed:
Merge branch 'beta' into keyStateManager

rueblimaster committed:
detekt

rueblimaster committed:
final touches

rueblimaster committed:
detekt

DavidArthurCole reviewed (changes requested):
There is a 0% chance we're touching this file again

rueblimaster commented:
> There is a 0% chance we're touching this file again

I heard David wants to durability test the new system. Go ahead @DavidArthurCole

DavidArthurCole commented:
> > There is a 0% chance we're touching this file again
> 
> 
> 
> I heard David wants to durability test the new system. Go ahead @DavidArthurCole 

Yeah sure give me til 2040

rueblimaster committed:
Merge branch 'beta' into keyStateManager

rueblimaster committed:
Merge branch 'beta' into keyStateManager

rueblimaster committed:
Merge branch 'beta' into KeyBindUtils

rueblimaster committed:
Merge branch 'KeyBindUtils' into keyStateManager

rueblimaster committed:
improve KeyHeldEvent posts

rueblimaster committed:
Revert "Merge branch 'KeyBindUtils' into keyStateManager"

This reverts commit 00239c367c70725d6ead7e676bf4d8572a5f0f10, reversing
changes made to a6c665a135667595eb4ffff58c97515d763a592e.

rueblimaster committed:
KeyHeldEvent also with mouse

rueblimaster committed:
imports

rueblimaster committed:
Merge branch 'beta' into keyStateManager

rueblimaster committed:
Merge branch 'beta' into keyStateManager

rueblimaster committed:
exclude from 1.21

rueblimaster committed:
Create baseline.xml

rueblimaster committed:
formatting

rueblimaster committed:
Merge branch 'beta' into keyStateManager

rueblimaster committed:
Merge branch 'beta' into keyStateManager

rueblimaster committed:
Merge branch 'beta' into keyStateManager

rueblimaster committed:
Merge branch 'beta' into keyStateManager

rueblimaster committed:
Merge branch 'beta' into keyStateManager

rueblimaster committed:
Merge branch 'beta' into keyStateManager

rueblimaster committed:
Merge branch 'beta' into keyStateManager

rueblimaster committed:
Merge branch 'beta' into keyStateManager

rueblimaster committed:
Merge branch 'beta' into keyStateManager

rueblimaster committed:
Remove KeyUpEvent and update key event references

Deleted the KeyUpEvent class and replaced references to KeyDownEvent with KeyPressEvent in DungeonSpiritLeapOverlay and KeyboardManager. This streamlines key event handling and removes unused code.

rueblimaster committed:
Replace KeyDown/KeyUp events with KeyHeld/KeyRelease in modern mixins

Updated keyboard and mouse event handling to use KeyHeldEvent and KeyReleaseEvent instead of KeyDownEvent and KeyUpEvent. This change standardizes event naming and improves consistency in input event processing.

DavidArthurCole commented on src/main/java/at/hannibal2/skyhanni/config/features/rift/RiftConfig.kt:
```diff
@@ -59,7 +59,7 @@ class RiftConfig {
     @ConfigOption(
         name = "Temporal Pillar Dodge",
         desc = "Avoid pathfinding solutions through or near the big enderman " +
-            "that throws you back to the end credits screen and steals so much time.",
```
Why is this here

DavidArthurCole commented on src/main/java/at/hannibal2/skyhanni/data/KeyStateManager.kt:
```diff
@@ -0,0 +1,85 @@
+
+        var button = Mouse.getEventButton()
+        if (button == -1) return // No button pressed
+        button = button - 100
```
Use Enums - magic numbers around keyboard/mouse inputs have caused, and will continue to cause extreme chaos until we stop using raw ints.

rueblimaster replied:
you're right this would probably be smarter, but as org.lwjgl.input.Mouse only gives us these as ints, I don't think that's a smart thing to do.
As the config always uses ints and also with enums we would limit the expandability, like what if someone with 8 mouse buttons came around someday

DavidArthurCole commented on src/main/java/at/hannibal2/skyhanni/events/minecraft/KeyHeldEvent.kt:
```diff
@@ -0,0 +1,6 @@
+import at.hannibal2.skyhanni.api.event.SkyHanniEvent
+
+/** Gets posted repeatedly while a key is held down, also on initial key press, use this for holding keys*/
+class KeyHeldEvent(val keyCode: Int) : SkyHanniEvent()
```
Convert this to a real Kdoc

DavidArthurCole commented on src/main/java/at/hannibal2/skyhanni/events/minecraft/KeyPressEvent.kt:
```diff
@@ -2,5 +2,5 @@ package at.hannibal2.skyhanni.events.minecraft
 
-/** Gets posted each tick it's pressed down*/
+/** Gets posted when a key is first pressed, use this for taps*/
 class KeyPressEvent(val keyCode: Int) : SkyHanniEvent()
```
Convert this to a real Kdoc

DavidArthurCole commented on src/main/java/at/hannibal2/skyhanni/events/minecraft/KeyReleaseEvent.kt:
```diff
@@ -0,0 +1,6 @@
+import at.hannibal2.skyhanni.api.event.SkyHanniEvent
+
+/** Gets posted when a key is released, counterpart to [KeyPressEvent]*/
+class KeyReleaseEvent(val keyCode: Int) : SkyHanniEvent()
```
Convert this to a real Kdoc

DavidArthurCole commented on src/main/java/at/hannibal2/skyhanni/data/KeyStateManager.kt:
```diff
@@ -0,0 +1,85 @@
+@SkyHanniModule
+object KeyStateManager {
+    // inventory check is not needed, as forge only send events when not in a gui
+    private fun shouldTrigger(): Boolean = !NeuItems.neuHasFocus()
```
Does this apply the same to Fabric? If we're trying to make something that's version agnostic, this feels like an oversight.

If this is _not_ being used with Fabric, and is solely for 1.8.9 use, rename this to `ForgeKeyStateManager`.

DavidArthurCole commented on src/main/java/at/hannibal2/skyhanni/utils/KeyboardManager.kt:
```diff
@@ -1,29 +1,19 @@
-import org.lwjgl.input.Mouse
 //#else
 //$$ import net.minecraft.client.util.InputUtil
 //#endif
```
Detekt will not like this.

DavidArthurCole commented on src/main/java/at/hannibal2/skyhanni/utils/KeyboardManager.kt:
```diff
@@ -224,7 +109,7 @@ object KeyboardManager {
-     * Do not use in KeyPressEvent, since it won't be unlocked again, use KeyDownEvent instead.
+     * Do not use in KeyPressEvent, since it won't be unlocked again, use KeyPressEvent instead.
      * */
     fun Int.isKeyClicked(): Boolean = if (this.isKeyHeld()) {
```
Kdoc, note what `this` is meant to be in the context.

DavidArthurCole commented on versions/1.21.5/src/main/java/at/hannibal2/skyhanni/mixins/transformers/MixinKeyboard.java:
```diff
@@ -41,13 +41,13 @@ private void onKey(long window, int key, int scancode, int action, int modifiers
+        if (action == 0) new KeyReleaseEvent(key).post();
         if (action == 1) {
-            new KeyDownEvent(key).post();
-            // on 1.21 it takes like 1 full second before the key press event will get posted so im doing it here
             new KeyPressEvent(key).post();
+            // on 1.21 it takes like 1 full second before the key press event will get posted so im doing it here
+            new KeyHeldEvent(key).post();
         }
-        if (action == 2) new KeyPressEvent(key).post();
+        if (action == 2) new KeyHeldEvent(key).post();
```
Again, use enums.

DavidArthurCole commented on versions/1.8.9/detekt/baseline.xml:
```diff
@@ -64,6 +64,28 @@
+    <ID>PreprocessingImportOrdering:LowHealthAlertSoundConfig.kt$//#if TODO</ID>
+    <ID>PreprocessingImportOrdering:RepositoryConfig.kt$//#if TODO</ID>
+    <ID>PreprocessingImportOrdering:SkyHanniMod.kt$//#if TODO</ID>
+    <ID>PreprocessingImportOrdering:TitleAndFooterConfig.kt$//#if TODO</ID>
```
No.

rueblimaster committed:
formatting

rueblimaster committed:
Format event class documentation comments

Updated the documentation comments in KeyHeldEvent, KeyPressEvent, and KeyReleaseEvent to use standard Kotlin KDoc block format for improved readability and consistency.

rueblimaster committed:
Refactor mouse button assignment in KeyStateManager

Replaces 'button = button - 100' with 'button -= 100' for improved code clarity and consistency.

rueblimaster committed:
Rename KeyStateManager to ForgeKeyStateManager

Renamed the KeyStateManager object and its file to ForgeKeyStateManager for improved clarity and consistency with Forge-specific functionality.

rueblimaster committed:
Update conditional import for MC 1.21+

Changed the preprocessor directive to import InputUtil only for Minecraft versions 1.21 and above.

rueblimaster committed:
Fix documentation for isKeyClicked usage

Updated the comment in KeyboardManager.kt to correctly reference KeyHeldEvent and KeyPressEvent, clarifying proper usage of isKeyClicked.

rueblimaster committed:
Update baseline.xml

rueblimaster committed:
Merge branch 'beta' into ForgeKeyEvents

rueblimaster committed:
Create baseline.xml

rueblimaster committed:
Merge branch 'beta' into ForgeKeyEvents

rueblimaster committed:
Merge branch 'beta' into ForgeKeyEvents

rueblimaster committed:
Merge branch 'beta' into ForgeKeyEvents

NopoTheGamer commented:
I think its just easier if we drop 1.8 support instead of worrying about any small issues in the 1.8 keybind system
